### DISCUSSION TITLE
tar_visnetwork has funny zoom settings

### DISCUSSION THREAD

**Discussion by busmansholiday** [ropensci/targets#925] [Help]
Running vis_networks in a standard targets project and get this image back:

Any suggestions on how to normalize this so that I can see nodes? 

![image](https://user-images.githubusercontent.com/108430289/189243404-f94ffb34-bea8-4467-8e45-5aab8f53fdc1.png)

**Comment by wlandau** [comment]
You could try a higher value for the `level_separation` argument. Or you could try temporarily commenting out some of the targets to get a more readable graph. You could comment out `tar_target()` calls or filter out a representative subset of rows for the `values` argument of `tar_map()`.

**Comment by dblodgett-usgs** [comment]
I have seen this behavior and it feels like something's wrong. 

`tar_visnetwork(targets_only = T)` is all I'm doing here. 

If I comment out some targets I can get a "nice" graph like:  

![image](https://github.com/ropensci/targets/assets/1492803/bdf4071b-fe3d-49fb-89f3-3d0c76b278ac)

If I add one more target I get:

![image](https://github.com/ropensci/targets/assets/1492803/712dfc02-fdc1-4e76-aa19-0d40525ea755)

level_seperation controls the lateral spacing but I'm not seeing how to control the vertical spacing that's getting blown out.

**Comment by dblodgett-usgs** [comment]
@wlandau -- I'm not up on R6 debugging, but I did confirm that by just returning `vertices` from `position_level` (https://github.com/ropensci/targets/blob/main/R/class_visnetwork.R#L185) I can get the vertical explosion to stop. I'll keep experimenting and maybe create a reprex for this. It's a tricky one to nail down.

**Reply by dblodgett-usgs** [reply]
OK, I _think_ this is actually an issue in visHierarchicalLayout. 

I extracted some data from the guts of the visnetwork class -- `out` here https://github.com/ropensci/targets/blob/main/R/class_visnetwork.R#L134 

One that causes the stretching behavior and one that does not. They are in the attached zip.

[visnetwork.zip](https://github.com/ropensci/targets/files/12833925/visnetwork.zip)

See the reprex below for what I'm seeing. For some reason, once "level" gets big, "level seperation" has to be small? Seems like there's some funky threshold behavior that's causing this to explode.

``` r
f <- file.path(tempdir(), "visnetwork.zip")
download.file("https://github.com/ropensci/targets/files/12833925/visnetwork.zip", f)
unzip(f, exdir = dirname(f))
f1 <- readRDS(file.path(dirname(f), "visnetwork_nostretch.rds"))
f2 <- readRDS(file.path(dirname(f), "visnetwork_stretch.rds"))

summary(f1$x$nodes$level)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>   1.000   2.000   3.000   3.788   6.000  10.000

visNetwork::visHierarchicalLayout(
  graph = f1,
  direction = "LR",
  levelSeparation = 150
)
```

![](https://i.imgur.com/NSF5ICt.png)

``` r

summary(f2$x$nodes$level)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>   1.000   2.000   5.000   6.977  11.500  20.000

visNetwork::visHierarchicalLayout(
  graph = f2,
  direction = "LR",
  levelSeparation = 150
)
```

![](https://i.imgur.com/KlbfYd0.png)

``` r

f3 <- f2
f3$x$nodes$level <- f3$x$nodes$level * 10

summary(f3$x$nodes$level)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>   10.00   20.00   50.00   69.77  115.00  200.00

visNetwork::visHierarchicalLayout(
  graph = f3,
  direction = "LR",
  levelSeparation = 10
)
```

![](https://i.imgur.com/uz8MlbA.png)

<sup>Created on 2023-10-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by dblodgett-usgs** [reply]
I found a fairly simple workaround for this. If we set physics to TRUE here: https://github.com/ropensci/targets/blob/main/R/class_visnetwork.R#L101 the stretched out graph compacts its self. Would you consider exposing that parameter in `tar_visnetwork()`?

**Reply by wlandau** [reply]
Sure, I think that argument makes sense to add. Implemented in eb2951df6e3c281459c6beef7ce8de80b4de9818.

**Reply by dblodgett-usgs** [reply]
Right on, Thanks @wlandau. For the record, I hacked this together for the time being. 

```r
#' @description 
#'   calls tar_visnetwork and manipulates the returned visNetwork object to fix up the layout when the aspect ratio is way out of whack.
#' @param physics logical passed to `visNetwork::visNodes` to modify the physics setting of the visNetwork object
#' @param scale numeric modifies the values of the node levels in the vixNetwork object.
#' @param levelSeperation numeric passed to `visNetwork::visHierarchicalLayout` to modify the levelSeperation parameter of the visNetwork object
my_visNetwork <- function(physics = FALSE, scale = 1, levelSeperation = 150) {
  
  net <- tar_visnetwork(targets_only = T, exclude = dplyr::ends_with("_file"))
  
  net <- visNetwork::visNodes(net, physics = physics)
  
  net$x$nodes$level <- net$x$nodes$level * scale
  
  net <- visNetwork::visHierarchicalLayout(net, levelSeparation = levelSeperation, direction = "LR")
 
  net 
}
```

**Reply by kaijagahm** [reply]
Has there been any more discussion about this behavior? It's happening a lot with my pipelines, and even though I can use `physics`, the resulting graph takes a long time (sometimes minutes) to finish compacting. This makes it really hard to monitor things at a glance (which is usually why I want to use visNetwork in the first place).
